Initial Concept of the alpineJS support

tsukhu opened the pull request:
- Passing parameters to `alpineJS`. In Alpine from what I can understand we need to have all the required props to be available in the` x-data` . So we need a mechanism to ensure that the `xData` attribute has the ability to receive props. And the plugin will process the `xData` and convert it to the `x-data` attribute with the `props` passed down and converted to a JSON string. Also if the user is not interested in the `props` from singlespa you can simply either *not set* xData and use the template or just pass your own initial state as an object.
- Coming to the `x-init` this will take a parameter that will be passed down by the plugin and will provide the `div id` or `app name` to get the correct `dom selector`. This can then be used to set the data using the `__x.$data` attributes. The way I have tried to implement this is that if the developer/user is providing the `x-init` then it must be provided with the `xData` as we need to have both `x-init` and `x-data` created on the same root div that the plugin will do. Also all we are doing is passing the `div id` to search for to the `xInit` function and its upto the user to take care of the function implementation.
- Now coming to the global function that needs to be created for each `xInit` , I have global variable `xInitFn` which will have keys as the `appName` which is the same as the `div id` . For each parcel which is declaring the `xInit` we will take the user defined function and add it globally to the `xInitFn` . 

- For example: 
```
const App3Data = ({ title, name }) => ({
  title,
  intro:
    'Implement a simple <code class="text-md text-pink-600">fetch()</code> request to render a list of items using Alpine.js :)',
  users: [],
  open: false,
  name,
});
const myFunc = (id) => {
  return fetch("https://jsonplaceholder.typicode.com/users")
    .then((response) => response.json())
    .then((data) => (document.querySelector(`#${id}`).__x.$data.users = data));
};
module.exports.mountApp3 = function mountApp3() {
  const opts = {
    template: templateApp3,
    xData: (data) => App3Data(data), // pass props to x-data
    xInit: myFunc,
  };
  const props = { name: "app3", title: "Alpine.js Landing Page" };
  return mount(opts, props);
}; 
```
- Here the `window.xInitFn = {"app3": myFunc}` and as we have more parcels more keys will be added .
- When we `unmount` that specific attribute will removed from `window.xInitFn` ensuring we are able to support multiple AlpineJS parcels

- So here is the summary
 The user can create Alpine Apps using the plugin in three ways
- `Only Template`
- `Template + xData` - where xData can be either an object or a function (not returning a promise - as x-data is not mention for async ops and we need to use x-init for that). If you provide a function , then we will push the single spa props down to the xData function and it will be initialized by the plugin with the props selected in xData
- `Template + xData + xInit` - where `xInit` can only exist when `xData` is defined as mentioned above. `xInit` has to be a function that will probably return a promise. It will also be injected with the `div id` for this app root element so that the user can take care of updating the `__x.$data` if required. This will ensure we don't get involved in this low level API should it change in the future

tsukhu committed:
added jshintrc for vscode

tsukhu committed:
first cut of the alpineJS plugin

internettrans commented on .jshintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "esversion": 9
```
We're using eslint, not jshint, which means that this file is probably not being used. Did you add it here for IDE support of something? I would prefer removing it since we're not using jshint.

tsukhu replied:
I was getting some IDE issues on VS Code earlier hence had added .. will remove

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -1,20 +1,46 @@
 
+  // xData can be optional , but when provided needs to be either a function or an object
+  if (
+    typeofXData !== "undefined" &&
```
This line disallows `xData` from being passed in as `null`. It's often confusing to treat null and undefined differently. A truthiness check likely makes the most sense, I think. One thing to note is taht `typeof null` is `"object"`, which causes problems here. The code below solves this:

```suggestion
    opts.xData &&
```

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -1,20 +1,46 @@
+  }
+
+  // xInit can be optional , but when provided needs to be a function
+  if (typeofXInit !== "undefined" && typeofXInit !== "function") {
```
Same here regarding truthiness check. Also, we don't need to extract `typeofXInit` as a variable, but rather can just inline it in the if statement. The reason I extracted `typeofTemplate` is because of line 44 where it's used outside of the validation logic.

```suggestion
  if (opts.xInit && typeof opts.xInit !== "function") {
```

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
-      // TODO - pass single-spa props to the alpinejs application somehow. Maybe through domElement.__x.$data.singleSpaProps?
+      // TODO - create any global x-init,x-data functions that are needed
+      // setup opts x-data value
+      if (typeofXData !== "undefined") {
```
A truthiness check here is better:

```suggestion
      if (opts.xData) {
```

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+        domElement.setAttributeNode(attXdata);
+
+        // Add x-init only if the x-data is provided as an object as both need to be on the same root dom element
+        if (typeofXInit !== "undefined") {
```
```suggestion
        if (opts.xInit) {
```

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+
+        if (typeofXData === "function") {
+          // TODO - pass single-spa props to the alpinejs application somehow. Maybe through domElement.__x.$data.singleSpaProps?
+          attXdata.value = JSON.stringify(opts.xData({ ...props })); // pass all singleSpaProps which will then are available to the xData function to consume
```
A couple ideas:

1. Perhaps we could just merge the opts.xData with the single-spa props to provide the final xData to alpine?

```js
const originalData = typeofXData === 'function' ? opts.xData({...props}) : opts.xData;
const finalData = Object.assign({}, props, originalData);
attXdata.value = JSON.stringify(finalData)
```

Thoughts?

2. Do you think it makes sense to allow the `xData` function to be asynchronous and return a Promise? Might be nice if someone wants to fetch it from server? Not sure if that's already handled in a different way inside of alpine.

```js
  // Returned promises from `.then()` functions will automatically flatten a returned promise into its raw value
  return typeofXData === 'function' ? opts.xData({...props}) : opts.xData;
}).then(originalData => {
})
```

tsukhu replied:
Point 1 - Great Idea and added to the implementation
Point 2 - I think this is a valid scenario, however I could not figure out how to implement it .

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+        if (typeofXInit !== "undefined") {
+          const appName = props.appName || props.name;
+          // create any global x-init functions that are needed
+          if (window.hasOwnProperty("xInitFn")) {
```
Checking my understanding on this - alpinejs requires that the function be global, but it can be named anything? So `xInitFn` could be any arbitrary name? I like what you've done here to support multiple applications being mounted, by adding them as properties to the object. However, I think it's a bit confusing that `window.xInitFn` is an object instead of a function. Perhaps rename to something like `window.singleSpaAlpineXInit`??? idk if that's the best name, but just something that doesn't call something a function when it's really an object

tsukhu replied:
Thats a great observation Joel. I will change it to singleSpaAlpineXInit

tsukhu replied:
done

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+          // create any global x-init functions that are needed
+          if (window.hasOwnProperty("xInitFn")) {
+            // add new x-init function globally for the specific ID
+            window.xInitFn = { ...window.xInitFn, [appName]: opts.xInit };
```
Immutability can be a nice pattern, but I don't see the benefit of using it here. This code is mutating `xInitFn` instead of mutating one of its properties - it is still mutation, and comes with a slight performance penalty.

```suggestion
            window.xInitFn[appName] = opts.xInit;
```

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+          attXdata.value = JSON.stringify(opts.xData);
+        }
+
+        domElement.setAttributeNode(attXdata);
```
Generally, single-spa helper libraries will not modify their container element, but rather append to it. The reason is that the container element is generally not destroyed/recreated when the application is mounted/unmounted/remounted.

The Vue library replaces the container element that you give it as a mount point, so we actually create a wrapper div to give to Vue so that it replaces that wrapper instead of replacing the domElement itself. See https://github.com/single-spa/single-spa-vue/blob/941c4a92f4e5919b4d1d28f4c166c38d600d4872/src/single-spa-vue.js#L96-L103.

In the case of single-spa-alpinejs, I'm okay with using the domElement directly **only if we remove any attributes during unmount**. Alternatively, we can create an inner container which has the xInit and xData attributes on it.

Another alternative we could explore is to add xInit/xData to `domElement.childElements[0]` after adding the template as innerHTML. That would require the template to only have one root element, though.

tsukhu replied:
Added a wrapper div 
```
      const domElement = domElementGetter(props);
      if (!domElement) {
        throw new Error(
          `single-spa-alpinejs: domElementGetter function for application '${
            props.appName || props.name
          }' did not return a valid dom element. Please pass a valid domElement or domElementGetter via opts or props`
        );
      }

      // create child Element
      const alpineDomElement = createAlpineElement(template, opts, props);

      // wrap in container div
      domElement.appendChild(alpineDomElement);
```

tsukhu replied:
The internal div will have an id `alpine-<appName>` and that will be used to mount and unmount the alpine code

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -98,6 +154,9 @@ function unmount(opts, props) {
 
     domElement.innerHTML = "";
     // TODO, remove any global functions that were used for x-init
+    if (window.hasOwnProperty("xInitFn")) {
```
We want single-spa-alpinejs' unmount function to die with a loud error if it cannot properly unmount. This is better than silent failure, since the user is aware of it and can fix it. The if statement on this line creates the possible of silent failures.

```suggestion
```

tsukhu replied:
Added a check for x-init
```
 // remove any global functions that were used for x-init
    if (opts.xInit) {
      if (
        !window.hasOwnProperty("xInitFn") ||
        typeof window.xInitFn[`${appName}`] === "undefined"
      ) {
        throw new Error(
          `single-spa-alpinejs: global function for xInit not found. Optional parameter opts.xInit if provided must be as a function that returns a promise`
        );
      }
      delete window.xInitFn[`${appName}`];
    }
```

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+      // TODO - create any global x-init,x-data functions that are needed
+      // setup opts x-data value
+      if (typeofXData !== "undefined") {
+        let attXdata = document.createAttribute("x-data"); // Add x-data attribute
```
Is there any advantage to using `createAttribute()` and `setAttributeNode` instead of calling the more familiar `setAttribute()`? I'd prefer using `setAttribute()` since it's more common and familiar.

tsukhu replied:
Agreed . It also reduces the lines of code ..

tsukhu replied:
done

internettrans commented on src/single-spa-alpinejs.test.js:
```diff
@@ -1,17 +1,269 @@
+    });
+    const domEl = domElementGetter();
+    return lifecycles
+      .bootstrap(props)
```
This test is great. Here's one tip with jest tests that I find handy - feel free to ignore/not change the code as there is no advantage to this approach over what you have. But you can make jest tests async functions and then use await for a bit of syntax sugar:

```js
it(`renders function template with x-data , template and props`, async () => {
  await lifecyles.bootstrap(props);
  await lifecycles.mount(props);
})
```

If you run into regeneratorRuntime issues, try setting the BABEL_ENV inside of the package.json:

https://github.com/single-spa/single-spa-alpinejs/blob/c7f720d151d4bedc3e589c3b54814f52d686fa54/package.json#L9

Example:

```json
{
  "scripts": {
    "test": "cross-env BABEL_ENV=test jest"
  }
}
```

See https://github.com/kentcdodds/cross-env and https://babeljs.io/docs/en/6.26.3/babelrc#env-option. Also https://github.com/single-spa/single-spa-alpinejs/blob/c7f720d151d4bedc3e589c3b54814f52d686fa54/.babelrc#L4

tsukhu replied:
Thats great insight Joel .. After adding some more tests I found that I was getting a Promise rejection error , so converted those tests to use async and they got resolved

tsukhu replied:
added this

internettrans reviewed (commented):
Great great work, thanks for your contribution here. I've left thorough review comments, but your work is high quality. Looking forward to getting this merged and published.

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -70,15 +98,43 @@ function mount(opts, props) {
+      if (typeofXData !== "undefined") {
+        let attXdata = document.createAttribute("x-data"); // Add x-data attribute
+
+        if (typeofXData === "function") {
```
```suggestion
        if (opts.xData) {
```

tsukhu replied:
done

tsukhu committed:
Update src/single-spa-alpinejs.js

Co-authored-by: Joel Denning <joeldenning@gmail.com>

tsukhu committed:
Update src/single-spa-alpinejs.js

Co-authored-by: Joel Denning <joeldenning@gmail.com>

tsukhu committed:
Update src/single-spa-alpinejs.js

Co-authored-by: Joel Denning <joeldenning@gmail.com>

tsukhu committed:
Update src/single-spa-alpinejs.js

Co-authored-by: Joel Denning <joeldenning@gmail.com>

tsukhu committed:
Update src/single-spa-alpinejs.js

agreed!

Co-authored-by: Joel Denning <joeldenning@gmail.com>

tsukhu committed:
fixes from code review

tsukhu committed:
post formatting

tsukhu committed:
improved tests and coverage

tsukhu committed:
tests and coverage additions

tsukhu commented:
> Great great work, thanks for your contribution here. I've left thorough review comments, but your work is high quality. Looking forward to getting this merged and published.

Thank you Joel for all your guidance

I have made all the changes suggested except one where we can support xData as a function returning a promise . If you could give me some tips on how to implement that , it will be great ! 

All other changes have been added .
Other major changes from yesterday
- There will be a DOM Wrapper div which will then have a child element for the alpine template , instead of modifying the root element and adding attributes of `x-data` and `x-init` there. Now we have an internal div created , that will be mounted and unmounted as well . For example: If the app name is `test` then the internal div will have an id `alpine-test` .

- I have added new tests to try and increase the test coverage which is ~75% now. There are some paths which may not be reached which are more of the safety checks of the `opts` which primarily get checked during the constructor and if they fail they will be catch there instead of in the `mount`
- Also added unmount api checks as part of the tests

internettrans commented on src/single-spa-alpinejs.js:
```diff
@@ -32,6 +53,49 @@ function bootstrap(opts, props) {
+  // create any global x-init,x-data functions that are needed
+  // setup opts x-data value
+  if (opts.xData) {
+    const originalData =
```
To support the `xData()` function returning promises, we can return the value from our `.then()` function and then add another `.then()` for when it finishes resolving. This works when returning both plain values and with promise values. The code below shows the idea

```js
export function mount(opts, props) {
  return Promise.resolve().then(() => {
    if (opts.xData) {
      return typeof opts.xData === 'function' ? opts.xData({...props}) : opts.xData
    } else {
      return {}
    }
  }).then(originalData => {
    const finalData = Object.assign({}, props, originalData)
  })
}
 
```

internettrans commented:
Looking good - let's make it possible for the xData function to return a promise and I think we'll be ready to merge. I have commented above on how to do so.

tsukhu committed:
added support for xData to be able to handle a function returning a promise

tsukhu commented:
> Looking good - let's make it possible for the xData function to return a promise and I think we'll be ready to merge. I have commented above on how to do so.

I have been able now understand and implement xData function handling of promises. Thank you Joel - this is a really a good learning as well. The updated code is now checked in

The new commit https://github.com/single-spa/single-spa-alpinejs/commit/290fde2751b7f25903ed749f9b2451bc2c79ea32

internettrans reviewed (approved):
Looks good to me. Thanks for your work on this. I will plan on publishing a 1.0.0 version to npm. We need to document this now.
